src/irmin.mli

```diff
@@ -1796,75 +1575,625 @@ module Private: sig
+
+  (** Repositories. *)
+  module Repo: sig
+
+    (** {1 Repositories}
+
+        A repository contains a set of branches. *)
+
+    type t
+    (** The type of repository handles. *)
+
+    val v: config -> t Lwt.t
+    (** [v mk_task config] connects to a repository in a
+        backend-specific manner. *)
+
+    val heads: t -> commit list Lwt.t
+    (** [heads] is {!Head.list}. *)
+
+    val branches: t -> branch list Lwt.t
+    (** [branches] is {Branch.list}. *)
+
+    val export: ?full:bool -> ?depth:int ->
+      ?min:commit list -> ?max:commit list ->
+      t -> slice Lwt.t
+    (** [export t ~depth ~min ~max] exports the store slice between
+        [min] and [max], using at most [depth] history depth (starting
+        from the max).
+
+        If [max] is not specified, use the current [heads]. If [min] is
+        not specified, use an unbound past (but can still be limited by
+        [depth]).
+
+        [depth] is used to limit the depth of the commit history. [None]
+        here means no limitation.
+
+        If [full] is set (default is true), the full graph, including the
+        commits, nodes and contents, is exported, otherwise it is the
+        commit history graph only. *)
+
+    val import: t -> slice -> [`Ok | `Error] Lwt.t
+    (** [import t s] imports the contents of the slice [s] in [t]. Does
+        not modify branches. *)
+
+    val task_of_commit: t -> commit -> task option Lwt.t
+    (** [task_of_commit t c] is the description of the commit
+        [c]. Useful to retrieve the commit date and the committer
+        name. Return [None] if [c] is not present in [t]. *)
+
+  end
+
+  val empty: Repo.t -> t Lwt.t
+  (** [empty repo task] is a temporary, empty store. Becomes a
+      normal temporary store after the first update. *)
+
+  val master: Repo.t -> t Lwt.t
+  (** [master repo] is a persistent store based on [r]'s master
+      branch. This operation is cheap, can be repeated multiple
+      times. *)
+
+  val of_branch: Repo.t -> branch -> t Lwt.t
+  (** [of_branch r name] is a persistent store based on the branch
+      [name]. Similar to [master], but use [name] instead
+      {!Branch.S.master}. *)
+
+  val of_commit: Repo.t -> commit -> t Lwt.t
+  (** [of_commit r c] is a temporary store, based on the commit
+      [c].
+
+      Temporary stores do not have stable names: instead they can be
+      addressed using the hash of the current commit. Temporary stores
+      are similar to Git's detached heads. In a temporary store, all
+      the operations are performed relative to the current head and
+      update operations can modify the current head: the current
+      stores's head will automatically become the new head obtained
+      after performing the update. *)
+
+  val repo: t -> Repo.t
+  (** [repo t] is the repository containing [t]. *)
+
+  val tree: t -> tree Lwt.t
+  (** [tree t] is [t]'s current tree. Contents is not allowed at the
+      root of the tree. *)
```

Comment:
I've tried to do that, but then it starts to be annoying to cast subtypes everywhere. There is probably a more elegant solution to this, but I haven't found it yet.